### ISSUE TITLE
`InstallBaseHandler` incorrect/incomplete paths

### ISSUE THREAD

**Issue by MatthijsBurgh** [ruffsl/colcon-clean#48]
The locations a package can be installed can be determined by the package itself. I think this is all defined in the `build_base/PKG_NAME/install_manifest.txt`. I think deleting the files from file is the most robust solution.

As the current implementation only deletes `install_base/PKG_NAME`. Though, this directory never exists.

Any solution based on patterns will be incomplete.

Looking forward to your opinion.

**Comment by ruffsl** [comment]
Ah, it looks like this is something provided by `ament_cmake`. Thanks! TDIL...

- https://github.com/ament/ament_cmake/pull/432

> I think this is all defined in the `build_base/PKG_NAME/install_manifest.txt`.

As a personal anecdote, looking through a workspace with Nav2, most `install_manifest.txt` files look empty? :thinking: 

<details>

```
ruffsl@c2970e878568:/opt/overlay_ws$ find ./build/nav2_* -name "install_manifest.txt" -exec sh -c 'echo {}; cat {}' \;
./build/nav2_amcl/install_manifest.txt
./build/nav2_behavior_tree/install_manifest.txt
./build/nav2_behaviors/install_manifest.txt
./build/nav2_bringup/install_manifest.txt
./build/nav2_bt_navigator/install_manifest.txt
./build/nav2_collision_monitor/install_manifest.txt
./build/nav2_common/install_manifest.txt
./build/nav2_constrained_smoother/install_manifest.txt
./build/nav2_controller/install_manifest.txt
./build/nav2_core/install_manifest.txt
./build/nav2_costmap_2d/install_manifest.txt
./build/nav2_dwb_controller/install_manifest.txt
./build/nav2_lifecycle_manager/install_manifest.txt
./build/nav2_map_server/install_manifest.txt
./build/nav2_mppi_controller/install_manifest.txt
./build/nav2_msgs/install_manifest.txt
/opt/overlay_ws/install/nav2_msgs/lib/libnav2_msgs__rosidl_generator_c.so
/opt/overlay_ws/install/nav2_msgs/lib/libnav2_msgs__rosidl_typesupport_fastrtps_c.so
/opt/overlay_ws/install/nav2_msgs/lib/libnav2_msgs__rosidl_typesupport_fastrtps_cpp.so
/opt/overlay_ws/install/nav2_msgs/lib/libnav2_msgs__rosidl_typesupport_introspection_c.so
/opt/overlay_ws/install/nav2_msgs/lib/libnav2_msgs__rosidl_typesupport_c.so
/opt/overlay_ws/install/nav2_msgs/lib/libnav2_msgs__rosidl_typesupport_introspection_cpp.so
/opt/overlay_ws/install/nav2_msgs/lib/libnav2_msgs__rosidl_typesupport_cpp.so
/opt/overlay_ws/install/nav2_msgs/lib/libnav2_msgs__rosidl_generator_py.so
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_generator_cExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_generator_cExport-release.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_typesupport_fastrtps_cExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_typesupport_fastrtps_cExport-release.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_generator_cppExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_typesupport_fastrtps_cppExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_typesupport_fastrtps_cppExport-release.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/nav2_msgs__rosidl_typesupport_introspection_cExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/nav2_msgs__rosidl_typesupport_introspection_cExport-release.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/nav2_msgs__rosidl_typesupport_cExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/nav2_msgs__rosidl_typesupport_cExport-release.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/nav2_msgs__rosidl_typesupport_introspection_cppExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/nav2_msgs__rosidl_typesupport_introspection_cppExport-release.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/nav2_msgs__rosidl_typesupport_cppExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/nav2_msgs__rosidl_typesupport_cppExport-release.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_generator_pyExport.cmake
/opt/overlay_ws/install/nav2_msgs/share/nav2_msgs/cmake/export_nav2_msgs__rosidl_generator_pyExport-release.cmake./build/nav2_navfn_planner/install_manifest.txt
./build/nav2_planner/install_manifest.txt
./build/nav2_regulated_pure_pursuit_controller/install_manifest.txt
./build/nav2_rotation_shim_controller/install_manifest.txt
./build/nav2_rviz_plugins/install_manifest.txt
/opt/overlay_ws/install/nav2_rviz_plugins/lib/libnav2_rviz_plugins.so
/opt/overlay_ws/install/nav2_rviz_plugins/share/nav2_rviz_plugins/cmake/nav2_rviz_pluginsExport.cmake
/opt/overlay_ws/install/nav2_rviz_plugins/share/nav2_rviz_plugins/cmake/nav2_rviz_pluginsExport-release.cmake./build/nav2_smac_planner/install_manifest.txt
./build/nav2_smoother/install_manifest.txt
./build/nav2_theta_star_planner/install_manifest.txt
./build/nav2_util/install_manifest.txt
./build/nav2_velocity_smoother/install_manifest.txt
./build/nav2_voxel_grid/install_manifest.txt
./build/nav2_waypoint_follower/install_manifest.txt
```

</details>

Aside from that, having to check a build base path in order to delete a package's install base path adds a conditional order of operations. E.g. if the build base path was cleaned prior and no longer exists, then we'd have no `install_manifest` to rely on. Any suggestions on how to handle this? That, and the case where a package provides a `install_manifest`, but never includes anything, or packages unrelated to cmake?

**Comment by MatthijsBurgh** [comment]
I really forgot about this issue.

I would suggest to go for the best we can do, without any guarantees. So I suggest to check the `install_manifest` file. As we are collecting paths, before deleting. It shouldn't be an issue we have a dependency on the build base. The question is, how do we get it correctly.

> That, and the case where a package provides a install_manifest, but never includes anything, or packages unrelated to cmake?

I don't understand what situation you are describing here. Anything that is in the `install_manifest` should be deleted IMO.

**Comment by ruffsl** [comment]
> what situation you are describing here

The one above when building Nav2.

**Comment by MatthijsBurgh** [comment]
For cases when the `install_manifest` is empty, we just don't have a better solution for now. Also for non-cmake packages, I also don't have solution ready yet.

Preferably all this package specific cleaning logic would be implemented in the plugins providing support for i.e. `ament_cmake` or `ament_python` packages. Though I don't know whether that is possible with the current plugin/extension framework.

**Comment by ruffsl** [comment]
> Though I don't know whether that is possible with the current plugin/extension framework.

With colcon's use of extension points, I think this would be very feasible. Colcon build extensions could expose extension points that encode how to clean after the package types they target, and colcon clean could infer what extension point to execute based on the detective package type and registered build extension.

I think that would be the most robust and generalizable approach, but perhaps ambitious. We could host the extension points in the colcon clean python module for now to simplify things in case the extension point function signatures need some design iterations. For example, this would be quite similar to how the base path extensions for install/build/log are handled already.

**Comment by MatthijsBurgh** [comment]
Are you okay with first merging, when finished, https://github.com/colcon/colcon-clean/pull/49, before we start working on these extension points?

How should we determine which extension point implementation to use for which pkg? Preferably this is checked by the implementation itself IMO.

**Comment by ruffsl** [comment]
> Are you okay with first merging, when finished, https://github.com/colcon/colcon-clean/pull/49, before we start working on these extension points?

With added tests, I would. Although would be nice to have an iteration on the extension points in before cutting a release.

> How should we determine which extension point implementation to use for which pkg? Preferably this is checked by the implementation itself IMO.

I think we could use package identification/augmentation to inspect what type of builder was used/prioritized for a package, as IIRC that is how registered builder extensions already work, so packages should already be identified by their respective build type.

- https://colcon.readthedocs.io/en/released/developer/extension-point.html#packageaugmentationextensionpoint

I did something like that before in `colcon-cache` to inspect what hash backend to use depending upon the presence or absence of version control for a given package's source:

- https://github.com/ruffsl/colcon-cache/blob/842696e3e437717046fd31a43a50be4070f23640/colcon_cache/package_augmentation/git.py#L35
- https://github.com/ruffsl/colcon-cache/blob/842696e3e437717046fd31a43a50be4070f23640/colcon_cache/package_augmentation/dirhash.py#L26C13-L26C21

> Perhaps I could have used identification extension there instead, but IIRC a package eventually resolves to the identity with the greatest priority, so I went with package augmentation to avoid interfering with the detected build type.

I.E augmentation would be after packages have been discovered and identified as cmake or python package:

- https://github.com/colcon/colcon-cmake/blob/921fe5c5381fdeeb7d22a3a257c91c6ac3e783de/colcon_cmake/package_identification/cmake.py#L26
- https://github.com/colcon/colcon-python-setup-py/blob/c78cbd60dec4ebf5284554a939f74d994962f61b/colcon_python_setup_py/package_identification/python_setup_py.py#L40

Once the build type for a package is determined, then the appropriate logic could be formulated to clean assets for that given build process and registered via `colcon-clean`'s extension points, that colcon-clean could then call out to when invoked upon a selected package.

Edit: Looking over the extension points again, I think we'd just need develop an endpoint that colcon build type extensions could use to register an ExecutorExtensionPoint used when processing colcon clean ~jobs (or ~tasks?).

Edit2: Ehh, then again, not sure. Could be something even more simple like preprocessor or function that returns a map or augments the file list. Not yet sure how to generalize the function signature or order of operations for integration.

**Comment by MatthijsBurgh** [comment]
Ok, I am bit busy at the moment. But I will try to get the PR ready before december and hopefully come with a design for the extension points etc. in december.
